rust/ares/src/codegen.rs

```diff
@@ -0,0 +1,1294 @@
+                    let a = slot(bend, 6).unwrap();
+                    let b = slot(bend, 14).unwrap();
+                    let v = slot(bend, 30).unwrap();
+                    let u = slot(bend, 62).unwrap().as_direct().unwrap().data() as usize;
+                    let n = slot(bend, 63).unwrap();
+                    let mut path = slot(n, 2).unwrap();
+                    let axis = slot(n, 3).unwrap().as_atom().unwrap();
+
+                    if let Some(jet) = context.hot.lookup(&mut context.stack, &mut path, axis) {
+                        let subject = register_get(current_frame, u);
+                        let jet_result = jet(context, subject);
+                        match jet_result {
+                            Ok(result) => {
+                                do_return(
+                                    context,
+                                    &mut current_frame,
+                                    virtual_frame,
+                                    &mut body,
+                                    &mut bend,
+                                    result,
+                                )
+                                .unwrap();
+                            }
+                            Err(JetErr::Fail(err)) => {
+                                ActualError(err);
+                            }
+                            Err(JetErr::Punt) => {
+                                do_tail_call(
+                                    context,
+                                    &mut current_frame,
+                                    &mut body,
+                                    &mut bend,
+                                    a,
+                                    b,
+                                    v,
+                                );
+                            }
+                        }
+                    } else {
+                        do_tail_call(context, &mut current_frame, &mut body, &mut bend, a, b, v);
+                    }
+                }
+                tas!(b"spy") => {
+                    let e = slot(bend, 6).unwrap().as_direct().unwrap().data() as usize;
+                    let p = slot(bend, 14).unwrap().as_direct().unwrap().data() as usize;
+                    let d = slot(bend, 30).unwrap().as_direct().unwrap().data() as usize;
+                    let mut t = slot(bend, 31).unwrap();
+
+                    if let Some(cell) = context.scry_stack.cell() {
+                        let scry_ref = register_get(current_frame, e);
+                        let scry_path = register_get(current_frame, p);
+                        let scry_stack = context.scry_stack;
+                        let scry_handler = cell.head();
+                        let scry_gate;
+                        if let Ok(cell) = scry_handler.as_cell() {
+                            scry_gate = cell;
+                        } else {
+                            return Err(ActualError(Error::ScryCrashed(D(0))));
```

Comment:
We shouldn't ever `return` directly from the loop. Just break from it. This allows easier debugging because we can insert debugging code after the loop exit
```suggestion
                            break Err(ActualError(Error::ScryCrashed(D(0))));
```